src/env.rs

```diff
@@ -260,10 +282,16 @@ impl Source for Environment {
                     }
                 } else {
                     // Skip this key
-                    return;
+                    return Ok(());
                 }
             }
 
+            // At this point, we don't know if key is required or not. Therefore the value is not a valid unicode
+            // string we error out.
+            let value = value
+                .into_string()
+                .map_err(|os_string| ConfigError::Foreign(InvalidUnicode(os_string).into()))?;
```

Comment:
Do we gain much benefit from creating a one-off error just for it to be errased when we can use `ConfigError::Message` like in #641?

Reply:
It allows the user to create specific handling for the error type. (I'll add the faulting key as well; so for example - remove the key from env and retry)

Reply:
While they can, will they? We can always migrate from few guarentees to more later but the other way is a breaking change. When we speculatively add features, it is also hard to know if people rely on them and why to then plan breaking changes.